platform/pulp/plugin/changeset/main.py

```diff
@@ -0,0 +1,200 @@
+        log.info(activity)
+        with Batch(DownloadIterator(self.adds, self.deferred)) as batch:
+            with ProgressBar(message=activity, total=len(self.adds)) as bar:
+                for plan in batch():
+                    artifact = plan.download.attachment
+                    artifact.model.downloaded = True
+                    content = artifact.content
+                    if not plan.succeeded:
+                        task = Task()
+                        task.append_non_fatal_error(plan.error)
+                        bar.increment()
+                        report = ChangeReport(ChangeReport.ADDED, content.model)
+                        report.errors.append(plan.error)
+                        yield report
+                        continue
+                    content.settle()
+                    if not content.settled:
+                        continue
+                    bar.increment()
+                    with transaction.atomic():
+                        self._update_catalog(content)
+                        self._associate(content.model)
+                    report = ChangeReport(ChangeReport.ADDED, content.model)
+                    yield report
+
+    def _update_catalog(self, content):
+        """
+        Update the content catalog.
+        Add (or update) an entry for each artifact.
+
+        Args:
+            content (RemoteContent): A newly added content.
+        """
+        for artifact in content.artifacts:
+            log.info(_('Cataloging: {a}').format(a=self))
+            try:
+                entry = DownloadCatalog()
+                entry.importer = self.importer
+                entry.url = artifact.download.url
+                entry.artifact = artifact.model
+                entry.save()
+            except IntegrityError:
+                entry = DownloadCatalog.objects.filter(
+                    importer=self.importer,
+                    artifact=artifact)
+                entry.url = artifact.download.url
+                entry.save()
+
+    def _remove_content(self):
+        """
+        Delete unwanted content.
+
+        Yields:
+            DeleteReport: A report for each removed content.
+        """
+        activity = _('Removing content.')
+        log.info(activity)
+
+        with ProgressBar(message=activity, total=len(self.deletes)) as bar:
+            for content in self.deletes:
+                self._disassociate(content)
+                bar.increment()
+                report = ChangeReport(ChangeReport.DELETED, content)
+                yield report
+
+    def apply(self):
+        """
+        Apply the changes to the repository based on download policy.
+
+        Yields:
+            ChangeReport: For each change applied.
+        """
+        log.info(_('Apply'))
+        for report in self._add_content():
+            yield report
+        for report in self._remove_content():
+            yield report
+
+
+class SizedIterable(Sized, Iterable):
```

Comment:
I understand your concern.  There will always be the potential for inconsistencies between the number of items yielded by an iterable and the length provided by the plugin writer.  I modeled as a `Sized`, `Iterable`  object explicitly because the combination of the interfaces places a high level of responsibility on the user (the plugin writer) to supply good data.  It formalizes the responsibility more than passing multiple related parameters to the ChangeSet.

I think that progress reporting is a best effort thing.  Nothing should ever be using progress reporting for logic or correctness.  It's information provided to the user so they can know that progress is being made and potentially make decisions like _"Will this finish before I go to lunch"_.  Given this perspective, I think logging an error may be useful but raising an error would be unproductive and not always possible.

I'm fine with making the `length` and and `iterable` protected/private to make the object immutable.